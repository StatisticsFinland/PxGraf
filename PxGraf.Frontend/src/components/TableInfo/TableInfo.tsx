import { Link } from "react-router-dom";
import { useTranslation } from 'react-i18next';
import { ListItemText, ListItemButton, Typography, Divider } from '@mui/material';
import React from 'react';
import { urls } from 'Router';
import { spacing } from 'utils/componentHelpers';
import UiLanguageContext from 'contexts/uiLanguageContext';
import { IDatabaseTable } from 'types/tableListItems';

interface ITableInfoProps {
    path: string;
    item: IDatabaseTable;
}

export const TableInfo: React.FC<ITableInfoProps> = ({ path, item }) => {
    const { t } = useTranslation();
    const { language } = React.useContext(UiLanguageContext);
    const currentPath = [path, item.fileName];
    const displayLanguage = item.languages.includes(language) ? language : item.languages[0];

    return (
        <>
            <ListItemButton id="mainContent" component={Link} to={urls.editor(currentPath)}>
                <ListItemText primary={
                    <>
                        <Typography variant="body1" sx={{ ...spacing(1) }}>
                            {item.name[displayLanguage]}
                        </Typography>
                        <Typography variant="body2" sx={{ ...spacing(1) }}>
                            {t("tableSelect.updated") + ": " + new Date(item.lastUpdated).toLocaleString(language)}
                        </Typography>
                    </>
                } />
            </ListItemButton>
            <Divider />
        </>
    );
}

export default TableInfo;