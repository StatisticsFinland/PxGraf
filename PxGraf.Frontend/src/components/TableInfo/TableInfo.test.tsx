import { render } from '@testing-library/react';
import { TableInfo } from './TableInfo';
import '@testing-library/jest-dom';
import UiLanguageContext from 'contexts/uiLanguageContext';
import { MemoryRouter } from 'react-router-dom';

const mockPath = 'asd123';
const mockItem = {
    id: 'id',
    updated: '2021-10-13T14:53:06',
    text: { 'fi': 'text-fi', 'en': 'text-en', 'sv': 'text-sv'},
    languages: ['fi', 'en', 'sv'],
};

const mockTableQueryResult = {
    data: {
        variables: [
            {
                code: 'code',
                text: 'text',
                valueTexts: 'valueTexts'
            }
        ]
    }
}

jest.mock('envVars', () => {
    return {
        PxGrafUrl: 'test-url.fi/',
    };
})

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

jest.mock('api/services/table', () => ({
    ...jest.requireActual('api/services/table'),
    useTableQuery: (currentPath: string, language: string) => {
        return mockTableQueryResult;
    },
}));

const setLanguage = jest.fn();
const language = 'fi';

const setLanguageTab = jest.fn();
const languageTab = 'fi';

const availableUiLanguages = ['fi', 'en', 'sv'];

const uiContentLanguage = "fi";
const setUiContentLanguage = jest.fn();

describe('Rendering test', () => {
    it('renders correctly', () => {
        // mock router for testing purposes
        const { asFragment } = render(<MemoryRouter><UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}><TableInfo path={mockPath} item={mockItem} /></UiLanguageContext.Provider></MemoryRouter>);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('displays text in the correct language when the language is included in item.languages', () => {
        const { getByText } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <TableInfo path={mockPath} item={mockItem} />
                </UiLanguageContext.Provider>
            </MemoryRouter>
        );

        expect(getByText('text-fi')).toBeInTheDocument();
    });

    it('displays text in the correct language when the language is not included in item.languages', () => {
        const { getByText } = render(
            <MemoryRouter>
                <UiLanguageContext.Provider value={{ language: 'fr', setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
                    <TableInfo path={mockPath} item={mockItem} />
                </UiLanguageContext.Provider>
            </MemoryRouter>
        );

        expect(getByText('text-fi')).toBeInTheDocument();
    });
});