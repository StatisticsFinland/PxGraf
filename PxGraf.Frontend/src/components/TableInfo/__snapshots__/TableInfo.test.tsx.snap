// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <a
    class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-4bx69l-MuiButtonBase-root-MuiListItemButton-root"
    data-discover="true"
    href="/editor/asd123/id/"
    id="mainContent"
    tabindex="0"
  >
    <div
      class="MuiListItemText-root css-cfq8qh-MuiListItemText-root"
    >
      <span
        class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-rizt0-MuiTypography-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-5arffz-MuiTypography-root"
        >
          text-fi
        </p>
        <p
          class="MuiTypography-root MuiTypography-body2 css-9f49mr-MuiTypography-root"
        >
          tableSelect.updated: 13.10.2021 klo 14.53.06
        </p>
      </span>
    </div>
  </a>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
  />
</DocumentFragment>
`;
